.github/workflows/ci.yml

```diff
@@ -125,35 +139,53 @@ jobs:
       name: Install graphviz
       if: ${{ matrix.install_graphviz }}
     # Add verbose flag to pip installation if in debug mode
-    - run: pip install -e .${{ matrix.extras }} ${{ fromJSON('["","-v"]')[runner.debug] }}
+    - run: pip install -e .${{ matrix.extras }} ${{ fromJSON('["","-v"]')[runner.debug] }} ${{ env.use_lkg && '-r lkg-notebook.txt' }}
       name: Install econml
-    - run: pip install pytest pytest-runner jupyter jupyter-client nbconvert nbformat seaborn xgboost tqdm
-      name: Install test and notebook requirements
-    - run: pip list
-      name: List installed packages
+    # Install notebook requirements (if not already done as part of lkg)
+    - run: pip install jupyter jupyter-client nbconvert nbformat seaborn xgboost tqdm
+      name: Install notebook requirements
+      if: ${{ !env.use_lkg }}
+    - run: pip freeze --exclude-editable > notebooks-${{ matrix.version }}-${{ matrix.kind }}-requirements.txt
+      name: Save installed packages
+    - uses: actions/upload-artifact@v3
+      name: Upload installed packages
+      with:
+        name: requirements
+        path: notebooks-${{ matrix.version }}-${{ matrix.kind }}-requirements.txt
+    - run: pip install pytest pytest-runner coverage
+      name: Install pytest
+
     - run: python setup.py pytest
       name: Run notebook tests
+      id: run_tests
       env:
         PYTEST_ADDOPTS: '-m "notebook"'
         NOTEBOOK_DIR_PATTERN: ${{ matrix.pattern }}
-    
+        COVERAGE_PROCESS_START: 'setup.cfg'
+    - run: mv .coverage .coverage.${{ matrix.kind }}
+      # Run whether or not the tests passed, but only if they ran at all
+      if: success() || failure() && contains(fromJSON('["success", "failure"]'), steps.run_tests.outcome)
+      name: Make coverage filename unique
+    - uses: actions/upload-artifact@v3
+      name: Upload coverage report
+      if: success() || failure() && contains(fromJSON('["success", "failure"]'), steps.run_tests.outcome)
+      with:
+        name: coverage
+        path: .coverage.${{ matrix.kind }}    
+
   tests:
     name: "Run tests"
     needs: [eval]
     if: ${{ needs.eval.outputs.testCode == 'True' }}
     strategy:
       matrix:
         os: [ubuntu-latest, windows-latest, macos-latest]
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: ['3.7', '3.8', '3.9', '3.10']
```

Comment:
Does changing the python versions to be strings do anything?

Reply:
They should always have been strings (and in many cases YAML interprets things as strings even without quotes, but since these can be parsed as floats that isn't true here); now it matters because `3.10` as a float is `3.1` which is not what we want as a version of python.

Reply:
There are a few other places in this file where versions are (incorrectly but benignly for now) specified as floats which I'll also change now.